### ISSUE TITLE
Starting in 0.24 kitty overrides fish autoloaded user-defined functions

### ISSUE THREAD

**Issue by mateuszwieloch** [kovidgoyal/kitty#4452]
**Describe the bug**
I noticed that something changed between kitty 0.23.1 and 0.24 that causes kitty to override fish user defined functions, e.g. `fish_prompt` and `fish_title`. This is a pretty bad default UX for fish users.

Fish functions autoloading is a feature of fish that loads scripts containing functions from `$fish_function_path` (by default it includes `~/.config/fish/functions` folder among others).

I see that the override that kitty makes happens here https://github.com/kovidgoyal/kitty/blob/master/shell-integration/fish/vendor_conf.d/kitty-shell-integration.fish#L42

I'm aware of one solution to this problem which is to modify kitty's `shell_integration` setting, but I wonder if it would make sense to have kitty detect that these functions were modified by a user and not override them.

P.S. Not sure if this issue should be classified as a bug (for existing users) or an enhancement.

**To Reproduce**
1. Run kitty 0.24
2. Run `type --all fish_prompt
3. Notice how it wasn't autoloaded by fish from ~/.config/fish/functions, but is "interactively defined".

**Screenshots**
N/A

**Environment details**
```
kitty 0.24.0 created by Kovid Goyal
Darwin MLC02CW63QMD6T 20.6.0 Darwin Kernel Version 20.6.0: Tue Oct 12 18:33:42 PDT 2021; root:xnu-7195.141.8~1/RELEASE_X86_64 x86_64
ProductName:	macOS ProductVersion:	11.6.1 BuildVersion:	20G224
Frozen: True
Paths:
  kitty: /Applications/kitty.app/Contents/MacOS/kitty
  base dir: /Applications/kitty.app/Contents/Resources/kitty
  extensions dir: /Applications/kitty.app/Contents/Resources/Python/lib/kitty-extensions
  system shell: /usr/local/bin/fish
Loaded config files:
  /Users/mwieloch/.config/kitty/kitty.conf

Config options different from defaults:
allow_remote_control     y
enabled_layouts          ['fat:bias=75', 'fat:bias=40', 'tall', 'stack']
font_family              FiraMono Nerd Font
font_size                14.0
kitty_mod                8
macos_custom_beam_cursor True
macos_option_as_alt      3
scrollback_lines         20000
tab_bar_min_tabs         1
Added shortcuts:
	cmd+. move_tab_forward
	cmd+[ previous_window
	cmd+] next_window
	cmd+` move_window_to_top
	cmd+a > 1 set_background_opacity 1
	cmd+a > d set_background_opacity default
	cmd+a > l set_background_opacity -0.1
	cmd+a > m set_background_opacity +0.1
	cmd+b move_window_backward
	cmd+e open_url_with_hints
	cmd+f show_scrollback
	cmd+g show_last_command_output
	cmd+h previous_tab
	cmd+j scroll_line_down
	cmd+k scroll_line_up
	cmd+l next_tab
	cmd+o pass_selection_to_program
	cmd+p > f kitten hints --type path --program -
	cmd+p > h kitten hints --type hash --program -
	cmd+p > l kitten hints --type line --program -
	cmd+p > n kitten hints --type linenum
	cmd+p > w kitten hints --type word --program -
	cmd+p > y kitten hints --type hyperlink
	cmd+p > shift+f kitten hints --type path
	cmd+q close_tab
	cmd+s paste_from_selection
	cmd+u kitten unicode_input
	cmd+x scroll_to_prompt 1
	cmd+z scroll_to_prompt -1
	cmd+escape kitty_shell window
	cmd+backspace change_font_size all 0
	cmd+delete clear_terminal reset active
	cmd+left previous_tab
	cmd+right next_tab
	cmd+f2 edit_config_file
	cmd+f5 load_config_file
	cmd+f6 debug_config
	cmd+f7 focus_visible_window
	cmd+f8 swap_with_window
	cmd+f10 toggle_maximized
	cmd+f11 toggle_fullscreen
	cmd+kp_subtract change_font_size all -2.0
	cmd+kp_add change_font_size all +2.0
	shift+cmd+h move_tab_backward
	shift+cmd+l move_tab_forward
	shift+cmd+m next_layout
	shift+cmd+n toggle_layout stack
	shift+cmd+r set_tab_title
	shift+cmd+t new_window_with_cwd
	alt+cmd+t set_tab_title
Removed shortcuts:
	shift+ctrl++ change_font_size all +2.0
	shift+ctrl+, move_tab_backward
	shift+ctrl+- change_font_size all -2.0
	shift+ctrl+. move_tab_forward
	shift+ctrl+0 tenth_window
	shift+ctrl+1 first_window
	shift+ctrl+2 second_window
	shift+ctrl+3 third_window
	shift+ctrl+4 fourth_window
	shift+ctrl+5 fifth_window
	shift+ctrl+6 sixth_window
	shift+ctrl+7 seventh_window
	shift+ctrl+8 eighth_window
	shift+ctrl+9 ninth_window
	shift+ctrl+= change_font_size all +2.0
	shift+ctrl+[ previous_window
	shift+ctrl+] next_window
	shift+ctrl+` move_window_to_top
	shift+ctrl+a > 1 set_background_opacity 1
	shift+ctrl+a > d set_background_opacity default
	shift+ctrl+a > l set_background_opacity -0.1
	shift+ctrl+a > m set_background_opacity +0.1
	shift+ctrl+b move_window_backward
	shift+ctrl+c copy_to_clipboard
	shift+ctrl+e open_url_with_hints
	shift+ctrl+f move_window_forward
	shift+ctrl+g show_last_command_output
	shift+ctrl+h show_scrollback
	shift+ctrl+j scroll_line_down
	shift+ctrl+k scroll_line_up
	shift+ctrl+l next_layout
	shift+ctrl+n new_os_window
	shift+ctrl+o pass_selection_to_program
	shift+ctrl+p > f kitten hints --type path --program -
	shift+ctrl+p > h kitten hints --type hash --program -
	shift+ctrl+p > l kitten hints --type line --program -
	shift+ctrl+p > n kitten hints --type linenum
	shift+ctrl+p > w kitten hints --type word --program -
	shift+ctrl+p > y kitten hints --type hyperlink
	shift+ctrl+p > shift+f kitten hints --type path
	shift+ctrl+q close_tab
	shift+ctrl+r start_resizing_window
	shift+ctrl+s paste_from_selection
	shift+ctrl+t new_tab
	shift+ctrl+u kitten unicode_input
	shift+ctrl+v paste_from_clipboard
	shift+ctrl+w close_window
	shift+ctrl+x scroll_to_prompt 1
	shift+ctrl+z scroll_to_prompt -1
	shift+ctrl+escape kitty_shell window
	shift+ctrl+enter new_window
	shift+ctrl+backspace change_font_size all 0
	shift+ctrl+delete clear_terminal reset active
	shift+ctrl+left previous_tab
	shift+ctrl+right next_tab
	shift+ctrl+up scroll_line_up
	shift+ctrl+down scroll_line_down
	shift+ctrl+page_up scroll_page_up
	shift+ctrl+page_down scroll_page_down
	shift+ctrl+home scroll_home
	shift+ctrl+end scroll_end
	shift+ctrl+f2 edit_config_file
	shift+ctrl+f5 load_config_file
	shift+ctrl+f6 debug_config
	shift+ctrl+f7 focus_visible_window
	shift+ctrl+f8 swap_with_window
	shift+ctrl+f10 toggle_maximized
	shift+ctrl+f11 toggle_fullscreen
	shift+ctrl+kp_subtract change_font_size all -2.0
	shift+ctrl+kp_add change_font_size all +2.0
	shift+alt+ctrl+t set_tab_title
Changed shortcuts:
	cmd+- decrease_font_size
	cmd+0 restore_font_size
	cmd+= increase_font_size
	cmd+t new_tab_with_cwd
	cmd+enter new_window_with_cwd
	shift+cmd+[ move_window_backward
	shift+cmd+] move_window_forward
	shift+cmd+d detach_window ask
	shift+cmd+w close_window

```

**Comment by page-down** [comment]
Fish shell functions can be defined in many places, and `fish_title` itself is `autoloaded` if you don't do any configuration.

```shell
functions --details --verbose fish_title
```

```text
/usr/local/share/fish/functions/fish_title.fish
autoloaded
1
scope-shadowing
```

So you should use the configuration option to turn off the features you don't need. Or disable it in the configuration and then manually configure the shell integration.

https://sw.kovidgoyal.net/kitty/shell-integration/

**Comment by kovidgoyal** [comment]
I dont know of a way to detect a "user defined" function vs one setup by the "system/a plugin/etc" if you do patches are most welcome.

**Comment by page-down** [comment]
Should the fish_title provided by the shell integration take precedence over the fish_title defined by the system, other plugins?

If ksi only overrides the fish_title that comes with fish, then I think I have a solution.

Otherwise it needs to define what is "user defined" first. Do user-installed plugins count as "user defined"?
I take a second look. There is almost no way to handle this very well.

**Comment by mateuszwieloch** [comment]
Thank you for looking into this Kovid! I really appreciate it.

I took a second look as well and I was surprised that both `$fish_function_path` and `$fish_user_paths` are not feasible for detecting a user-defined functions vs the default ones. And I couldn't find any other elegant way to do it. You're right that `shell_integration` is probably the best solution.

@page-down The ideal way I envisioned this to work is that if built-in functions are overridden by anything (plugin or user-defined function), kitty would not do its override. My thinking here is that kitty's override will be confusing for fish users.
